File: ThruHikerLogApi/ThruHikerLogApi/Services/TrailService.cs
using ThruHikerLogApi.Models;
using ThruHikerLogApi.Repos;

namespace ThruHikerLogApi.Services
{
    public class TrailService(TrailRepo trailRepo, EntryRepo entryRepo)
    {
        private readonly TrailRepo _trailRepo = trailRepo;
        private readonly EntryRepo _entryRepo = entryRepo;

        /// <summary>
        /// Create a new trail
        /// </summary>
        /// <param name="newTrail"></param>
        /// <param name="author"></param>
        public void CreateTrail(Trail newTrail, User? author)
        {
            _trailRepo.CreateTrail(newTrail, author);
        }

        /// <summary>
        /// Returns a hardcoded list of popular long-distance hiking trails in the United States.
        /// </summary>
        /// <returns>A list of Trails</returns>
        public IEnumerable<Trail> GetAllTrails()
        {
            return _trailRepo.GetTrails();
        }

        /// <summary>
        /// Gets the details for a specific trail, including its entries
        /// </summary>
        /// <param name="id"></param>
        /// <returns></returns>
        public Trail GetTrailDetails(int id)
        {
            var trail = _trailRepo.GetTrailDetails(id);
            var entries = _entryRepo.GetEntriesByTrailId(id);
            trail.Entries = entries.ToList();
            return trail;
        }

        /// <summary>
        /// Updates a trail
        /// </summary>
        /// <param name="id"></param>
        /// <param name="updatedTrail"></param>
        public void UpdateTrail(int id, Trail updatedTrail)
        {
            _trailRepo.UpdateTrail(id, updatedTrail);
        }

        /// <summary>
        /// Delete a trail and all associated entries
        /// </summary>
        /// <param name="id"></param>
        public void DeleteTrail(int id)
        {
            _trailRepo.DeleteTrail(id);
        }
    }
}

File: thru-hiker-log-app/src/app/services/trail.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ITrail } from '../models/trail';

@Injectable({ providedIn: 'root' })
export class TrailService {
  private baseUrl = 'https://localhost:7225/api/trail';

  constructor(private http: HttpClient) { }

  /** Create a trail */
  createTrail(trail: ITrail): Observable<void> {
    return this.http.post<void>(this.baseUrl, trail);
  }

  /** Get list of all trails */
  getAllTrails(): Observable<ITrail[]> {
    return this.http.get<ITrail[]>(this.baseUrl);
  }

  /** Get details for a specific trail */
  getTrailDetails(id: number): Observable<ITrail> {
    return this.http.get<ITrail>(this.baseUrl + '/' + id);
  }

  /** Update a trail */
  updateTrail(trail: ITrail): Observable<void> {
    return this.http.put<void>(this.baseUrl + '/' + trail.id, trail);
  }
}
